docs/architecture/psa-migration/md-cipher-dispatch.md

```diff
@@ -346,8 +346,39 @@ The primary target is a configuration like TF-M's medium profile, plus TLS with
 This excludes things like:
 - Support for encrypted PEM, PKCS5 and PKCS12 encryption, and PKCS8 encrypted keys in PK parse. (Not widely used on highly constrained devices.)
 - Support for NIST-KW. (Same justification.)
+- Support for CMAC. (Same justification, plus can be directly accelerated.)
 - Support for CBC ciphersuites in TLS. (They've been recommended against for a while now.)
 
+### Dual-dispatch for block cipher primitives
+
+Considering the priorities stated above, initially we want to support GCM, CCM and CTR-DRBG. All trhee of them use the block cipher primitive only in the encrypt direction. Currently, GCM and CCM use the Cipher layer in order to work with AES, Aria and Camellia (DES is excluded by the standards due to its smaller block size) and CTR-DRBG directly uses the low-level API from `aes.h`. In all cases, access to the "block cipher primitive" is done by using "ECB mode" (which for both Cipher and `aes.h` only allows a single block, contrary to PSA which implements actual ECB mode).
+
+The two AEAD modes, GCM and CCM, have very similar needs and positions in the stack, strongly suggesting using the same design for both. On the other hand, there are a number of differences between CTR-DRBG and them.
+- CTR-DRBG only uses AES (and there is no plan to extend it to other block ciphers at the moment), while GCM and CCM need to work with 3 block ciphers already.
+- CTR-DRBG holds a special position in the stack: most users don't care about it per se, they only care about getting random numbers - in fact PSA users don't even need to know what DRBG is used. In particular, no part of the stack is asking questions like "is CTR-DRBG-AES available?" - an RNG needs to be available and that's it - contrary to similar questions about AES-GCM etc. which are asked for example by TLS.
+
+So, it makes sense to use different designs for CTR-DRBG on one hand, and GCM/CCM on the other hand:
+- CTR-DRBG can just check if `AES_C` is present and "fall back" to PSA is not.
```

Comment:
Even in 3.6, we should try to prefer PSA if both are available, for builds that have software AES for legacy or bootstrap purposes only and would prefer to use PSA normally. But I'd put that as a nice to have, not as a core requirement.

Note that the AES key changes often during the lifetime of the DRBG, so it would make sense for a DRBG to start using legacy AES, then use PSA AES when that becomes available, with the decision being taken when setting the key.

Reply:
This would come at a cost in terms of code size: if we decide where to dispatch at runtime, it means we must have both code paths (PSA and legacy) compiled in. So, it's a trade-off.